internal/controlplane/common.go

```diff
@@ -36,44 +36,113 @@ type HasProtoContext interface {
 	ctx context.Context,
 	store db.Store,
 	in HasProtoContext,
 	projectId uuid.UUID,
 ) (db.Provider, error) {
+	name := getNameFilterParam(in.GetContext())
+	providers, err := findProvider(ctx, name, db.NullProviderType{}, projectId, store)
+	if err != nil {
+		return db.Provider{}, err
+	}
+
+	p, err := inferProvider(providers, name)
+	if err != nil {
+		return db.Provider{}, err
+	}
+
+	return p, nil
+}
+
+func getProvidersByTrait(
+	ctx context.Context,
+	store db.Store,
+	in HasProtoContext,
+	projectId uuid.UUID,
+	trait db.ProviderType,
+) ([]db.Provider, error) {
+	name := getNameFilterParam(in.GetContext())
+	t := db.NullProviderType{ProviderType: trait, Valid: true}
+	providers, err := findProvider(ctx, name, t, projectId, store)
+	if err != nil {
+		return nil, err
+	}
+
+	return providers, nil
+}
+
+// findProvider is a helper function to find a provider by name and trait
+func findProvider(
+	ctx context.Context,
+	name sql.NullString,
+	trait db.NullProviderType,
+	projectId uuid.UUID,
+	store db.Store,
+) ([]db.Provider, error) {
 	// Allows us to take into account the hierarchy to find the provider
 	parents, err := store.GetParentProjects(ctx, projectId)
 	if err != nil {
-		return db.Provider{}, status.Errorf(codes.InvalidArgument, "cannot retrieve parent projects: %s", err)
+		return nil, status.Errorf(codes.InvalidArgument, "cannot retrieve parent projects: %s", err)
 	}
-	providers, err := store.ListProvidersByProjectID(ctx, parents)
+
+	provs, err := store.FindProviders(ctx, db.FindProvidersParams{
+		Projects: parents,
+		Name:     name,
+		Trait:    trait,
+	})
 	if err != nil {
-		return db.Provider{}, status.Errorf(codes.InvalidArgument, "cannot retrieve providers: %s", err)
+		return nil, fmt.Errorf("cannot retrieve providers: %w", err)
 	}
-	// if we do not have a provider name, check if we can infer it
-	if in.GetContext().GetProvider() == "" {
+
+	if len(provs) == 0 {
+		return nil, filteredResultNotFoundError(name, trait)
+	}
+
+	return provs, nil
+}
+
+// getNameFilterParam allows us to build a name filter for our provider queries
+func getNameFilterParam(in *pb.Context) sql.NullString {
+	if in.GetProvider() == "" {
+		return sql.NullString{}
+	}
+	return sql.NullString{
+		String: in.GetProvider(),
+		Valid:  true,
+	}
+}
+
+func inferProvider(providers []db.Provider, nameFilter sql.NullString) (db.Provider, error) {
+	if !nameFilter.Valid {
 		if len(providers) == 1 {
 			return providers[0], nil
 		}
 		return db.Provider{}, util.UserVisibleError(codes.InvalidArgument, "cannot infer provider, there are %d providers available",
 			len(providers))
 	}
 
-	matchesName := func(provider db.Provider) bool {
-		return provider.Name == in.GetContext().GetProvider()
-	}
-
-	i := slices.IndexFunc(providers, matchesName)
-	if i == -1 {
-		return db.Provider{}, util.UserVisibleError(codes.InvalidArgument, "invalid provider name: %s", in.GetContext().GetProvider())
-	}
-	return providers[i], nil
+	return providers[0], nil
```

Comment:
I'm not sure I understand the logic here -- when `nameFilter` is valid, we'll return the first item, even if several were found?

What if zero rows were found, will this panic?  (I realize that right now, we only feed it results from `findProvider` which won't return a 0-length or `nil` array, but I worry about the function being re-used somewhere.  I think we could safely fold it into `getProviderFromRequestOrDefault`.

Reply:
this is not called if zero are found

Reply:
https://github.com/stacklok/minder/pull/2812